search_similar_entities/facetedSearchEndpoint.rb

```diff
@@ -0,0 +1,161 @@
+
+# Verify params are OK
+raise "Missing required parameter: id" unless params[:id]
+raise "Missing required parameter: mode - Values: [queryString|queryStringWithFacets]" unless params[:mode]
+raise "Acceptable values for parameter mode: [queryString|queryStringWithFacets]" unless "queryString|queryStringWithFacets|".include? params["mode"] + '|'
+thisId = params[:id]
+return_mode = params[:mode]
+
+## Utility Methods
+
+# Utility method to return the different "leaf" values found in the hierarchical attribute 
+# 'hierachicalAttr1' belonging
+# to a testEntity, that begin with level1value|level2value
+def getLeafValues(testEntity, level1value, level2value)
+  return testEntity['hierarchicalAttr1'].select { |a| a.start_with?(level1value + '|' + level2value) }.uniq
+end
+
+# Utility method that adds n and k elements to a hash
+def addNKElementsToHash(theHash, nObj, kObj)
+  unless nObj.nil? or nObj.empty?
+    theHash['n'] = nObj
+  end
+  unless kObj.nil? or kObj.empty?
+    theHash['k'] = kObj
+  end
+  return theHash
+end
+  
+
+# Utility method to return the data structure that goes in the queryURL (as JSON) 
+# to make sure a given refinement widget is selected for a given value
+def getSimpleRefWidgetDataStruct(widgetName,xPath,logic,theValue)
+  
+  theValueHash = {"n" => theValue}
+  tmpHash = {"id" => widgetName, "logic" => logic, "s" => [theValueHash]}
+  unless xPath.empty?
+    tmpHash['xPath'] = xPath
+  end
+  return tmpHash
+  
+end 
+
+# Utility method to return the data structure that goes in the queryURL (as JSON) 
+# to make sure a given refinement widget (that has hierarchical values) is selected for a given value
+def getComplexRefWidgetDataStruct(widgetName,xPath,logic,theValue,separator)
+  
+  tmpHash = {"id" => widgetName, "logic" => logic, "s" => theValue}
+  unless xPath.empty?
+    tmpHash['xPath'] = xPath
+  end
+  unless separator.empty?
+    tmpHash['separator'] = separator
+  end
+  return tmpHash
+  
+end 
+
+# Utility method to return the complex data structure that goes in the queryURL for a
+# nested refinement widget
+# theLabels has a number of labels separated by separator
+# starting from backwards, it adds the label, and the full path to the label
+def getComplexRefWidgetDataStructure(theLabels, separator)
+  labelArray = theLabels.split('|')
+  arrayOfHashes = []
+  until labelArray.empty? do
+    allLabels = labelArray.join('|')
+    h = {}
+    h = addNKElementsToHash(h,labelArray.last, allLabels)
+    arrayOfHashes << h
+    labelArray.pop # Remove the last label
+  end
+  # Now iterate over all the building blocks we've got so far, nesting each element as an "s" in the next hast
+  for i in 0 .. arrayOfHashes.count - 2
```

Comment:
Do you mean the until loop? I'm not sure how to do that with a map, as each iteration modifies the original array.
If it's the for loop, note that it leaves the last element out of the loop. A map would apply to all elements, wouldn't it?